Honda Bosch Radarless: experimental longitudinal support

royjr opened the pull request:
Possible pull for #25355. Works great except stops way too close to the lead car.
[See panda PR](https://github.com/commaai/panda/pull/1008)

sshane commented:
Can you provide some routes with rlogs uploaded in stop and go traffic and highway speeds? I'd like to confirm we understand the the accel interface and it follows our request well enough

sshane commented on selfdrive/car/honda/interface.py:
```diff
@@ -36,12 +36,7 @@ def get_params(candidate, fingerprint=gen_empty_fingerprint(), car_fw=[], disabl
-        # WARNING: THIS DISABLES AEB!
-        ret.openpilotLongitudinalControl = disable_radar
-
+      ret.openpilotLongitudinalControl = disable_radar
```
```suggestion
      # Disable the radar and let openpilot control longitudinal
      # WARNING: THIS DISABLES AEB!
      # If Bosch radarless, this blocks ACC messages
      ret.openpilotLongitudinalControl = disable_radar
```

sshane commented:
Also should figure out what sends AEB and FCW so we can make sure those are never blocked (I think the camera does it?)

royjr commented:
I assume FCW is the warning before AEB takes over and starts braking. I believe the camera sends those messges from our last [findings](https://github.com/commaai/opendbc/pull/651). I was never able to trigger AEB on my car confirm the AEB messages (AEB_BRAKING, AEB_STATUS)

sshane commented on selfdrive/car/honda/hondacan.py:
```diff
@@ -51,33 +51,40 @@ def create_acc_commands(packer, enabled, active, accel, gas, stopping, car_finge
+  if car_fingerprint in HONDA_BOSCH_RADARLESS:
+    acc_control_values = {
+      "CONTROL_ON": control_on,
+      "CONTROL_OFF": control_off,
```
Did you figure out which bit controls actuation, if any? Or do both?

royjr replied:
It looks like you only need CONTROL_ON just like the comment says. Tried without setting either CONTROL and with just CONTROL_OFF, but none moved the car or changed ACC_STATUS.

<img width="350" alt="1a5d045d2c531a6d|2022-08-08--18-41-25" src="https://user-images.githubusercontent.com/1976269/183527702-6c8a6ed4-9b64-4a94-a1b9-404073f76e28.png">
1a5d045d2c531a6d|2022-08-08--18-41-25

sshane commented on selfdrive/car/honda/hondacan.py:
```diff
@@ -51,33 +51,40 @@ def create_acc_commands(packer, enabled, active, accel, gas, stopping, car_finge
+      "ACCEL_COMMAND": accel_command,  # used for brakes
+      "BRAKE_LIGHTS": braking,
+      "BRAKE_REQUEST": braking,
+      "STANDSTILL": standstill,
```
It might be good to find any standstill bits in the message so we can set those if needed.

sshane commented:
- [ ] Figure out what the stock system requests to hold a standstill, and does it set any bits?

royjr committed:
civic22_long

sshane committed:
Stock requests -4

sshane committed:
add standstill bit

royjr commented:
1a5d045d2c531a6d|2022-09-01--13-46-28
1a5d045d2c531a6d|2022-09-01--13-36-53

royjr commented:
tux: ```undefined signal STANDSTILL - 456```
There is no STANDSTILL in ACC_CONTROL for this car.

sshane commented:
Should be fixed now

sshane commented:
Should be fixed now. Can you tell me a few things?

- With stock ACC, do you need to press the resume button or press the accelerator to resume from a stop?
- Before the STANDSTILL addition on openpilot ACC, do you need to press the resume button or press the accelerator to resume from a stop?

royjr commented:
**Stock ACC** - _needs resume button_
**OP + Stock ACC** - _does not need resume button, but sometimes doesnt resume automatically (once in a blue)_
**OP + OP ACC** - _does not need resume button, but short terms tests may not be conclusive based on above_
**OP + OP ACC + STANDSTILL addition** - _does not need resume button, but short terms tests may not be conclusive based on above_

royjr committed:
bump

royjr committed:
fix standstill

royjr committed:
bump

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
bump

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
bump panda

royjr committed:
Update hondacan.py

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
bump panda

royjr committed:
add docs

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
bump panda

royjr committed:
bump panda

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
fix control_on value

royjr committed:
set control_off

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
fix cars.md

royjr committed:
Update panda

royjr committed:
Update opendbc

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Update panda

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Update panda

royjr committed:
allow more braking

royjr committed:
Revert "allow more braking"

This reverts commit ecd2493a436a095b621d815e8b0c7591fa6b93a9.

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Update panda

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr commented:
@dmdreon It definitely works, just has a few ui kinks left to iron out.
- Mismatching speed when setting speed below 25
- Dashboard blinks speed sometimes
- Dashboard "stopped" remains after resume

royjr committed:
fix speed mismatch

royjr committed:
Revert "fix speed mismatch"

This reverts commit ca795dc5bd9d6e841225e1ac1ff93a1f78618513.

royjr committed:
Update panda

royjr committed:
fix speed mismatch and standstill

royjr committed:
delay stopped message to match stock behavior

royjr committed:
green cruise only when enabled

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Update panda

sshane commented on selfdrive/car/honda/hondacan.py:
```diff
@@ -111,13 +119,17 @@ def create_ui_commands(packer, CP, enabled, pcm_speed, hud, is_metric, acc_hud,
       'SET_ME_X01_2': 1,
     }
 
+    if CP.carFingerprint in HONDA_BOSCH_RADARLESS:
+      acc_hud_values['ENABLE_MINI_CAR'] = 1 if enabled else 0
+    else:
+      acc_hud_values['ENABLE_MINI_CAR'] = 1
+
```
this looks like a general improvement/clean up for all Honda. can we make it another PR? is it critical to this PR?

sshane commented on selfdrive/car/honda/hondacan.py:
```diff
@@ -58,27 +58,35 @@ def create_acc_commands(packer, enabled, active, accel, gas, stopping, car_finge
-
+  if car_fingerprint in HONDA_BOSCH_RADARLESS:
+    control_on = 1 if enabled else 0
+    acc_control_values = {
+      "CONTROL_ON": control_on,
+      "CONTROL_OFF": 1,  # TODO: not sure this signal is needed
+      "ACCEL_COMMAND": accel_command,
+      "STANDSTILL": standstill,
```
the signals for this message name defined already below, are they common at all in the new acc control msg? are there weird constant setmes?

royjr replied:
Some are common, but might have different values like CONTROL_ON. I moved common messages out of the if statement.

royjr committed:
Revert "green cruise only when enabled"

This reverts commit d21aa5b541a42fa6d1c87e3b065f720af28ff38b.

royjr committed:
lesser nester

royjr commented on selfdrive/car/honda/carcontroller.py:
```diff
@@ -127,7 +128,27 @@ def __init__(self, dbc_name, CP, VM):
+          if CS.is_metric:
+            hud_v_cruise = hud_control.setSpeed * CV.MS_TO_KPH if hud_control.speedVisible else 255
+          else:
+            hud_v_cruise = hud_control.setSpeed * CV.MS_TO_MPH if hud_control.speedVisible else 255
```
Slightly better?
```
if self.CP.carFingerprint in HONDA_BOSCH_RADARLESS:
  if CC.longActive and (actuators.longControlState == LongCtrlState.stopping):
    if self.last_stopping_frame == 0:
      self.last_stopping_frame = self.frame
    if self.frame - self.last_stopping_frame > 400:
      hud_v_cruise = 252
    else:
      hud_v_cruise = hud_control.setSpeed * (CV.MS_TO_KPH if CS.is_metric else CV.MS_TO_MPH) if hud_control.speedVisible else 255
  else:
    self.last_stopping_frame = 0
    hud_v_cruise = hud_control.setSpeed * (CV.MS_TO_KPH if CS.is_metric else CV.MS_TO_MPH) if hud_control.speedVisible else 255
else:
  hud_v_cruise = hud_control.setSpeed * CV.MS_TO_MPH if hud_control.speedVisible else 255
```

royjr replied:
Maybe even better?
```
if self.CP.carFingerprint in HONDA_BOSCH_RADARLESS:
  hud_v_cruise = hud_control.setSpeed * (CV.MS_TO_KPH if CS.is_metric else CV.MS_TO_MPH) if hud_control.speedVisible else 255
  if CC.longActive and (actuators.longControlState == LongCtrlState.stopping):
    if self.last_stopping_frame == 0:
      self.last_stopping_frame = self.frame
    if self.frame - self.last_stopping_frame > 400:
      hud_v_cruise = 252
  else:
    self.last_stopping_frame = 0
else:
  hud_v_cruise = hud_control.setSpeed * CV.MS_TO_MPH if hud_control.speedVisible else 255
```

royjr committed:
simplify

royjr committed:
simplify control_on

royjr committed:
simplify common signals

royjr committed:
simplify CONTROL_ON

royjr committed:
lessest nester

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
hide speed on gas_override to match stock behavior

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

sshane committed:
tiny simplification, lesser nester

sshane commented on selfdrive/car/honda/carcontroller.py:
```diff
@@ -127,7 +128,18 @@ def __init__(self, dbc_name, CP, VM):
+      if CC.longActive and actuators.longControlState == LongCtrlState.stopping:
+        self.stopping_counter += 1
+        if self.stopping_counter > 400:
+          hud_v_cruise = 252
```
can you make a global variable with a comment for this magic number of what it does?

royjr replied:
I'm assuming the magic number in reference is 400 and not 252, or should we make both variables?

sshane replied:
400 is very clearly 4 seconds, I meant what that set speed did

sshane replied:
Does showing stopped change any behavior like turn off the engine?

royjr replied:
Silly me, 400 wasn't 4 seconds to me because it would start counting down before it came to a stop. 252 is the stopped message which we now don't need.

sshane committed:
flip common acc control values around

royjr committed:
move magic number to global variable

royjr committed:
Revert "move magic number to global variable"

This reverts commit 160be0e0872dc08a8786d97fc2125b2eb131362d.

royjr committed:
remove stopped logic

royjr commented:
A few things come to mind..

- There was no conversion logic before, it was always CV.MS_TO_KPH which doesn't work for the civic. This might break non-radarless cars. Should we only apply the conversion to radarless?
- Do we still need the gas_override in the hondacan ui function or can we just do that check in carcontroller since both could set to the same 255 value?

royjr committed:
remove unused variable

sshane commented:
Yes, how you had it before with the radar less check was fine. We can bring out the override check too

royjr committed:
looks ok

royjr committed:
looks better

royjr committed:
simplify override

royjr commented:
done, unless you wanted to move the override check to another PR since thats new to other hondas

royjr committed:
clarify CONTROL_OFF signal comment

royjr committed:
civic22_long_bettercontroloff

shorten delay

init

selfless

whoops

move to carcontroller

selfmore

lower time, currently take 5s to kick in

just right

deline

royjr committed:
Merge pull request #1 from royjr/civic22_long_bettercontroloff

civic22_long_bettercontroloff

royjr committed:
Rename CONTROL_OFF -> IDLESTOP_ALLOW

This tells the car its ready to turn off the engine, which does so when the car is standstill.

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Update panda

royjr committed:
Update opendbc

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Update panda

royjr committed:
fix docs

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Update panda

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Update panda

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Update opendbc

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Update panda

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Update panda

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Update panda

sshane committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

royjr committed:
Update panda

sshane committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

sshane committed:
bump panda

sshane committed:
idlestop in another PR

sshane committed:
radarless doesn't need tester present

sshane committed:
simplify that

sshane committed:
helper

sshane committed:
one line is fine

sshane commented:
@royjr when this is merged, can you open new PRs for the follow?:

- idlestop timer, if stock uses a timer and doesn't set it immediately or statically
- match stock hide speed behavior on override

sshane committed:
Merge remote-tracking branch 'upstream/master' into civic22_long

sshane committed:
test bump

sshane committed:
bumpback

sshane commented on selfdrive/car/honda/hondacan.py:
```diff
@@ -51,34 +56,42 @@ def create_acc_commands(packer, enabled, active, accel, gas, stopping, car_finge
+  if car_fingerprint in HONDA_BOSCH_RADARLESS:
+    acc_control_values.update({
+      "CONTROL_ON": enabled,
+      "IDLESTOP_ALLOW": 1,  # allows idle stop
```
```suggestion
      "IDLESTOP_ALLOW": 0,  # disallows idle stop
```

sshane committed:
disallow

sshane commented on selfdrive/car/honda/hondacan.py:
```diff
@@ -51,34 +56,42 @@ def create_acc_commands(packer, enabled, active, accel, gas, stopping, car_finge
   min_gas_accel = CarControllerParams.BOSCH_GAS_LOOKUP_BP[0]
 
-  control_on = 5 if enabled else 0
   gas_command = gas if active and accel > min_gas_accel else -30000
```
```suggestion
  control_on = 5 if enabled else 0
  gas_command = gas if active and accel > min_gas_accel else -30000
```

sshane committed:
reduce diff
